File: golang/day01/day01.go
package main

import (
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"slices"
)

type problemData struct {
	left, right []int
}

func diffInt(a, b int) int {
	var diff int = a - b
	if diff < 0 {
		return 0 - diff
	}
	return diff
}

func readData(inputFile string) problemData {
	file, err := os.Open(inputFile)
	if err != nil {
		log.Fatal(err)
		os.Exit(1)
	}
	defer file.Close()

	var left, right int
	var data problemData

	for {
		_, err := fmt.Fscanln(file, &left, &right)

		if err == io.EOF {
			break
		}
		if err != nil {
			panic(err)
		}

		data.left = append(data.left, left)
		data.right = append(data.right, right)
	}

	return data
}

func runPart1(data problemData) int {
	left := make([]int, len(data.left))
	right := make([]int, len(data.right))

	copy(left, data.left)
	slices.Sort(left)

	copy(right, data.right)
	slices.Sort(right)

	difference := 0

	for i := range left {
		difference += diffInt(left[i], right[i])
	}

	return difference
}

func runPart2(data problemData) int {
	score := 0

	for _, left := range data.left {
		// Inefficient but not a speed-critical puzzle today
		count := 0
		for _, right := range data.right {
			if left == right {
				count++
			}
		}
		score += left * count
	}

	return score
}

func main() {
	var inputFile = flag.String("input", "inputs/day01.txt", "Problem input file")
	flag.Parse()
	data := readData(*inputFile)
	part1 := runPart1(data)
	part2 := runPart2(data)
	fmt.Printf("%d, %d\n", part1, part2)
}

File: python/day01.py
#!/usr/bin/env python3
'''Advent of Code 2024 Day 1 solution'''
import argparse
import copy

def read_data(filename):
    '''Read input data'''
    left_list = []
    right_list = []

    with open(filename, 'r', encoding="utf8") as file:
        for line in file:
            line_data = line.split()
            left_list.append(int(line_data[0]))
            right_list.append(int(line_data[1]))

    return(left_list, right_list)

def runpart1(data):
    '''Run part one'''
    left = data[0]
    left.sort()

    right = data[1]
    right.sort()

    difference = 0

    for index, l in enumerate(left):
        difference += abs(l - right[index])

    return difference

def runpart2(data):
    '''Run part two'''
    score = 0

    for i in data[0]:
        # Inefficient but not a speed-critical puzzle today
        score += i * len([j for j in data[1] if i == j])

    return score

def main():
    '''Main'''
    parser = argparse.ArgumentParser()
    parser.add_argument("filename", nargs='?', default="inputs/day01.txt")
    args = parser.parse_args()

    # Read input file
    data = read_data(args.filename)

    # Run the solution
    part1 = runpart1(copy.deepcopy(data))
    part2 = runpart2(data)
    return (part1, part2)

if __name__ == '__main__':
    print(main())
